hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/FileGroupReaderBasedMergeHandle.java

```diff
@@ -251,6 +344,31 @@ public List<WriteStatus> close() {
     }
   }
 
+  private Option<BaseFileUpdateCallback<T>> createCallback() {
+    List<BaseFileUpdateCallback<T>> callbacks = new ArrayList<>();
+    // Handle CDC workflow.
+    if (cdcLogger.isPresent()) {
+      callbacks.add(new CDCCallback<>(cdcLogger.get(), readerContext));
+    }
+    // Indexes are not updated during compaction
+    if (operation.isEmpty()) {
+      // record index callback
+      if (this.writeStatus.isTrackingSuccessfulWrites()) {
+        writeStatus.manuallyTrackSuccess();
+        callbacks.add(new RecordLevelIndexCallback<>(writeStatus, newRecordLocation, partitionPath));
+      }
+      // Stream secondary index stats.
+      if (isSecondaryIndexStatsStreamingWritesEnabled) {
+        callbacks.add(new SecondaryIndexCallback<>(
```

Comment:
I thought we had a discussion to have just 1 callback for RLI and SI. did you attempt and had to revert due to some issues?

Reply:
No, that is not what we discussed. The discussion was to avoid conversion to HoodieRecord just to update SI.